contribute/design/1.x/cstor/pdb/20200113-cstor-poddisruptionbudget.md

```diff
@@ -0,0 +1,170 @@
+---
+oep-number: Pod Disruption Budget
+title: Pod Disruption Budget for cStor
+authors:
+  - "@mittachaitu"
+owners:
+  - "@vishnuitta"
+  - "@kmova"
+  - "@AmitKumarDas"
+editor: "@mittachaitu"
+creation-date: 2020-01-13
+last-updated: 2019-01-13
+status: Implementable
+---
+
+# Pod Disruption Budget
+
+
+## Table of Contents
+
+- [Pod Disruption Budget](#pod-disruption-budget)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [HighAvailable Volumes Shouldn't be disturbed for Cluster Upgrade Activities](#highavailable-volumes-shouldnt-be-disturbed-for-cluster-upgrade-activities)
+    - [Proposed Implementation](#proposed-implementation)
+    - [Steps to perform user stories](#steps-to-perform-user-stories)
+    - [Low Level Design](#low-level-design)
+      - [Work Flow](#work-flow)
+      - [Sample Yaml](#sample-yaml)
+      - [Notes:](#notes)
+    - [TestCases](#testcases)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+  - [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This proposal includes design details of Pod Disruption Budget(PDB) for cStor pool
+pods.
+- Not to allow more than quorum no.of HighAvailable(HA i.e volume provisioned
+  with >= 3 replicas) volume replicas pool pods to go down.
+
+## Motivation
+
+There are cases arising where OpenEBS users upgrade cluster or take out the
+multiple nodes for maintenance due to this HA volume is going to ReadOnly mode
+without informing users/admin[How? Multiple pool pod nodes are taken at time].
+
+### Goals
+
+- Create Pod Disruption Budget among the cStor pool pods where HA volume replicas exists.
+  Below are examples of valid voluntary disruptions
+  - Draining a node for repair or upgrade.
+  - Draining a node from a cluster to scale the cluster down.
+
+### Non-Goals
+
+- Invalid/Other valid voluntary disruptions are not supported via PDB.
+
+## Proposal
+
+### User Stories
+
+#### HighAvailable Volumes Shouldn't be disturbed for Cluster Upgrade Activities
+As an OpenEBS user, HA volumes shouldn't be disturbed for cluster upgrade or maintenance
+activities.
+
+### Proposed Implementation
+
+Kubernetes will send volume creation request to CSI driver when PVC is created.
+CSI driver will send a cStorVolume provision request via CVC CR to cStorVolumeConfig(CVC)
+controller. CVC controller will schedule the replicas among the pools after that
+CVC controller will create a PDB among the pool pods where the volume replicas
+are scheduled.
+
+Example:
+- Provisioned a CSPC with five pool specs that intern creates 5 cStor Pools.
+- Now, when 'n' no.of HA volumes are provisioned(with replica count as 3) then
+  volume replicas can be schedule in any available cStor pools.
+- After volume replicas are placed then cStorVolumeConfig controller will
```

Comment:
PDB will be created only if PDB of those pools is not yet available.

Reply:
If PDB already exists, an label will be added on PDB with the PV name

Reply:
ok.. got to see that, if PDB already exists, a label will be added on CVC with the PDB name

Reply:
Added